code/network/multi_portfwd.cpp

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright (C) Volition, Inc. 1999.  All rights reserved.
+ *
+ * All source code herein is the property of Volition, Inc. You may not sell
+ * or otherwise commercially exploit the source or things you created based on the
+ * source.
+ *
+*/
+
+#ifndef _WIN32
+#include <arpa/inet.h>
+#endif
+
+#define STATICLIB	// required to link properly on windows
+#include "natpmp.h"
+
+#include "cmdline/cmdline.h"
+#include "io/timer.h"
+#include "network/multi_log.h"
+#include "network/psnet2.h"
+
+
+enum pf_type_t {
+	PF_TYPE_NATPMP		= 0,
+	PF_TYPE_UPNP		= 1
+};
+
+static pf_type_t PF_type = PF_TYPE_NATPMP;
+
+enum pf_status_t {
+	PF_STATUS_IDLE				= 0,
+	PF_STATUS_PENDING			= 1,
+	PF_STATUS_PUBLIC_IP_GET		= 2,
+	PF_STATUS_PUBLIC_IP_WAIT	= 3,
+	PF_STATUS_MAPPING_SET		= 4,
+	PF_STATUS_MAPPING_WAIT		= 5
+};
+
+static pf_status_t PF_status = PF_STATUS_PENDING;
+
+static int PF_wait_timestamp = 0;
+static int PF_renewal_timestamp = 0;
+static uint16_t PF_public_port = 0;
+static const uint32_t PF_lifetime = 7200;	// 2 hrs
+
+// NAT-PMP stuff
+static natpmp_t PF_natpmp;
+
+static void PF_natpmp_err(const int err, const bool cleanup = true);
+
+static bool PF_natpmp_init();
+static bool PF_natpmp_close();
+static void PF_natpmp_do();
+
+
+void multi_port_forward_init()
+{
+	bool rval;
+	bool renew = false;
+
+	if ( timestamp_elapsed(PF_renewal_timestamp) ) {
```

Comment:
The `timestamp_` functions are only usable while within a mission. Is that intentional? If so then the timestamp should at least be reset in `multi_port_forward_close` to make sure that the renewal is done properly. If that is not intentional then you could use the `timer_` functions instead to get a system timestamp.

Reply:
Thanks! 17 years working in this code and I'm still messing that up.